src/uct/ugni/smsg/ugni_smsg_ep.c

```diff
@@ -0,0 +1,268 @@
+                                uint8_t am_id, unsigned header_length, void *header,
+                                unsigned payload_length, void *payload, uct_ugni_smsg_desc_t *desc)
+{
+  gni_return_t gni_rc;
+
+  desc->msg_id = iface->smsg_id++;
+  desc->ep = &ep->super;
+
+  gni_rc = GNI_SmsgSendWTag(ep->super.ep, header, header_length, payload, payload_length, desc->msg_id, am_id);
+
+  if(gni_rc != GNI_RC_SUCCESS){
+    /* look for better return code. Maybe steak pasha's macro */
+    ucs_mpool_put(desc);
+    return UCS_ERR_NO_RESOURCE;
+  }
+
+  ++ep->super.outstanding;
+  ++iface->super.outstanding;
+
+  sglib_hashed_uct_ugni_smsg_desc_t_add(iface->smsg_list, desc);
+
+  return payload_length;
+}
+
+ucs_status_t uct_ugni_smsg_ep_am_short(uct_ep_h tl_ep, uint8_t id, uint64_t header,
+                                       const void *payload, unsigned length)
+{
+
+  uct_ugni_smsg_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_ugni_smsg_iface_t);
+  uct_ugni_smsg_ep_t *ep = ucs_derived_of(tl_ep, uct_ugni_smsg_ep_t);
+  uct_ugni_smsg_header_t *smsg_header;
+  uint64_t *header_data;
+  uct_ugni_smsg_desc_t *desc;
+  ucs_status_t rc;
+
+  UCT_CHECK_AM_ID(id);
+  UCT_CHECK_LENGTH(length, iface->config.smsg_seg_size -
+                   (sizeof(smsg_header) + sizeof(header)), "am_short");
+
+  UCT_TL_IFACE_GET_TX_DESC(&iface->super.super, &iface->free_desc,
+                           desc, return UCS_ERR_NO_RESOURCE);
+
+  ucs_trace_data("AM_SHORT [%p] am_id: %d buf=%p length=%u",
+                 iface, id, payload, length);
+
+  smsg_header = (uct_ugni_smsg_header_t *)(desc+1);
+  smsg_header->length = length + sizeof(header);
+
+  header_data = (uint64_t*)(smsg_header+1);
+  *header_data = header;
+  memcpy((void*)(header_data+1), payload, length);
+
+  uct_iface_trace_am(&iface->super.super, UCT_AM_TRACE_TYPE_SEND,
+                     id, header_data, length, "TX: AM_SHORT");
+
+  rc = uct_ugni_smsg_ep_am_common_send(ep, iface, id, sizeof(uct_ugni_smsg_header_t),
+                                       smsg_header, smsg_header->length, (void*)header_data, desc);
+
+  if(rc > 0){
+    return UCS_OK;
+  } else {
+    return rc;
+  }
+}
+
+ssize_t uct_ugni_smsg_ep_am_bcopy(uct_ep_h tl_ep, uint8_t id,
+                                  uct_pack_callback_t pack_cb,
+                                  void *arg)
+{
+  uct_ugni_smsg_iface_t *iface = ucs_derived_of(tl_ep->iface, uct_ugni_smsg_iface_t);
+  uct_ugni_smsg_ep_t *ep = ucs_derived_of(tl_ep, uct_ugni_smsg_ep_t);
+  ssize_t packed;
+  uct_ugni_smsg_desc_t *desc;
+
+  UCT_CHECK_AM_ID(id);
+
+  UCT_TL_IFACE_GET_TX_DESC(&iface->super.super, &iface->free_desc,
+                           desc, return UCS_ERR_NO_RESOURCE);
+
+  ucs_trace_data("AM_BCOPY [%p] am_id: %d buf=%p",
+                 iface, id, arg );
+
+  uct_ugni_smsg_header_t *smsg_header = (uct_ugni_smsg_header_t *)(desc+1);
+  void *smsg_data = (void*)(smsg_header+1);
```

Comment:
this is pure styling comment, but I don't like declarations in the middle of the code